src/bench/connectblock.cpp

```diff
@@ -0,0 +1,133 @@
+        std::vector{coinbase_to_spend},
+        std::vector{COutPoint(coinbase_to_spend->GetHash(), 0)},
+        chainstate.m_chain.Height() + 1, keys, outputs, std::nullopt, std::nullopt)};
+    const auto test_block_parent_coinbase{GetScriptForDestination(coinbase_taproot)};
+    test_setup->CreateAndProcessBlock(std::vector{first_tx}, test_block_parent_coinbase, &chainstate);
+
+    std::vector<CMutableTransaction> txs;
+    txs.reserve(num_txs);
+    CTransactionRef input_tx{MakeTransactionRef(first_tx)};
+    for (int i{0}; i < num_txs; i++) {
+        std::vector<COutPoint> inputs;
+        inputs.reserve(outputs.size());
+
+        for (size_t j{0}; j < outputs.size(); j++) {
+            inputs.emplace_back(input_tx->GetHash(), j);
+        }
+
+        const auto [taproot_tx, _]{test_setup->CreateValidTransaction(
+            std::vector{input_tx}, inputs, chainstate.m_chain.Height() + 1, keys, outputs, std::nullopt, std::nullopt)};
+        txs.emplace_back(taproot_tx);
+        input_tx = MakeTransactionRef(taproot_tx);
+    }
+
+    // Coinbase output can use any output type as it is not spent and will not change the benchmark
+    const CScript coinbase_spk{GetScriptForDestination(coinbase_taproot)};
+    return test_setup->CreateBlock(txs, coinbase_spk, chainstate);
+}
+
+std::pair<std::vector<CKey>, std::vector<CScript>> CreateKeysAndScripts(const CKey& coinbaseKey, size_t num_taproot, size_t num_nontaproot)
+{
+    std::vector<CKey> keys{coinbaseKey};
+    keys.reserve(num_taproot + num_nontaproot + 1);
+
+    std::vector<CScript> spks;
+    spks.reserve(num_taproot + num_nontaproot);
+
+    for (size_t i{0}; i < num_nontaproot; i++) {
+        const CKey key{GenerateRandomKey()};
+        keys.emplace_back(key);
+        spks.emplace_back(GetScriptForDestination(WitnessV0KeyHash{key.GetPubKey()}));
+    }
+
+    for (size_t i{0}; i < num_taproot; i++) {
+        CKey key{GenerateRandomKey()};
+        keys.emplace_back(key);
+        spks.emplace_back(GetScriptForDestination(WitnessV1Taproot{XOnlyPubKey(key.GetPubKey())}));
+    }
+
+    return {keys, spks};
+}
+
+void BenchmarkConnectBlock(benchmark::Bench& bench, std::vector<CKey>& keys, std::vector<CScript>& spks, std::shared_ptr<TestChain100Setup> test_setup)
+{
+    const auto test_block{CreateTestBlock(test_setup, keys, spks)};
+    auto pindex{std::make_unique<CBlockIndex>(test_block)};
+    auto test_blockhash{std::make_unique<uint256>(test_block.GetHash())};
+
+    Chainstate& chainstate{test_setup->m_node.chainman->ActiveChainstate()};
+
+    pindex->nHeight = chainstate.m_chain.Height() + 1;
+    pindex->phashBlock = test_blockhash.get();
+    pindex->pprev = chainstate.m_chain.Tip();
+
+    BlockValidationState test_block_state;
+    bench.unit("block").run([&] {
+        LOCK(cs_main);
+        CCoinsViewCache viewNew{&chainstate.CoinsTip()};
+        assert(chainstate.ConnectBlock(test_block, test_block_state, pindex.get(), viewNew, false));
```

Comment:
`fJustCheck` is already false by default, so I think this would be better as:

```suggestion
        assert(chainstate.ConnectBlock(test_block, test_block_state, pindex.get(), viewNew));
```